app/models/instance.rb

```diff
@@ -15,6 +15,9 @@ def find_tenant_by_host(host)
     end
   end
 
+  VALID_HOST_NAME_REGEX = /\A[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\z/i
+  validates :host, format: { with: VALID_HOST_NAME_REGEX }, length: { maximum: 255 }
```

Comment:
What label do you call the root? I don't know. I'm guessing. But trailing periods make sense to me, I've seen them in use before so I've no reason to believe they're invalid.

What you're looking for is probably normalisation, not validation.